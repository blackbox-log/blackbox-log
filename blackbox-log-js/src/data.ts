import { Memoize as memoize } from 'typescript-memoize';

import { WasmPointer } from './wasm';

import type { Headers } from './headers';
import type { WasmExports, WasmObject } from './wasm';

export class DataParser implements WasmObject {
	readonly #wasm: WasmExports;
	readonly #ptr: WasmPointer;
	readonly #headers: Headers;

	constructor(wasm: WasmExports, headers: Headers, ptr: number) {
		this.#wasm = wasm;
		this.#headers = headers;
		this.#ptr = new WasmPointer(ptr, wasm.data_free);
	}

	free() {
		this.#ptr.free();
	}

	get isAlive(): boolean {
		return this.#ptr.isAlive;
	}

	get headers(): Headers {
		return this.#headers;
	}

	parseAll() {
		this.#wasm.data_parseAll(this.#ptr.ptr);
	}

	get mainFrameCount(): number {
		return this.#wasm.data_mainFrameCount(this.#ptr.ptr);
	}

	get gpsFrameCount(): number {
		return this.#wasm.data_gpsFrameCount(this.#ptr.ptr);
	}
}
