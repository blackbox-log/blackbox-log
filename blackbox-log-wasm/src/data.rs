use blackbox_log::prelude::*;
use blackbox_log::Reader;

use crate::{OwnedSlice, Shared, WasmFfi};

// SAFETY: field order *must* be `parser` first, then `headers`, then `data` to
// ensure correct drop order
pub struct WasmDataParser {
    parser: DataParser<'static, 'static>,
    _headers: Shared<Headers<'static>>,
    _data: Shared<OwnedSlice>,
}

impl WasmDataParser {
    pub(crate) fn new(
        headers: Shared<Headers<'static>>,
        reader: Reader<'static>,
        data: Shared<OwnedSlice>,
    ) -> Self {
        // SAFETY: this is only used to create the `DataParser`, which is guaranteed to
        // be dropped before `headers` by the declaration order in the struct
        let headers_ref = unsafe { headers.deref_static() };

        Self {
            parser: DataParser::new(reader, headers_ref),
            _headers: headers,
            _data: data,
        }
    }

    fn main_frame_count(&self) -> usize {
        self.parser.stats().counts.main
    }

    fn gps_frame_count(&self) -> usize {
        self.parser.stats().counts.gps
    }
}

impl WasmFfi for WasmDataParser {}

#[no_mangle]
unsafe extern "wasm" fn data_free(ptr: *mut WasmDataParser) {
    let parser = WasmDataParser::from_wasm(ptr);
    drop(parser);
}

#[no_mangle]
#[allow(non_snake_case)]
unsafe extern "wasm" fn data_mainFrameCount(ptr: *mut WasmDataParser) -> usize {
    let parser = WasmDataParser::from_wasm(ptr);
    let count = parser.main_frame_count();
    parser.into_wasm();
    count
}

#[no_mangle]
#[allow(non_snake_case)]
unsafe extern "wasm" fn data_gpsFrameCount(ptr: *mut WasmDataParser) -> usize {
    let parser = WasmDataParser::from_wasm(ptr);
    let count = parser.gps_frame_count();
    parser.into_wasm();
    count
}
