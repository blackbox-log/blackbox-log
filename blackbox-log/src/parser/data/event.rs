use alloc::vec::Vec;

use tracing::instrument;

use crate::parser::{decode, ParseError, ParseResult, Reader};

#[derive(Debug, Clone, PartialEq)]
pub enum Event {
    SyncBeep(u64),
    InflightAdjustment {
        function: u8,
        new_value: AdjustedValue,
    },
    Resume {
        log_iteration: u32,
        time: u32,
    },
    Disarm(u32),
    FlightMode {
        flags: u32,
        last_flags: u32,
    },
    ImuFailure {
        error: u32,
    },
    End,
}

#[derive(Debug, Clone, Copy, PartialEq)]
pub enum AdjustedValue {
    Float(f32),
    Int(i32),
}

impl Event {
    #[instrument(level = "debug", name = "Event::parse", skip_all, fields(kind))]
    pub(crate) fn parse_into(data: &mut Reader, events: &mut Vec<Self>) -> ParseResult<EventKind> {
        let byte = data.read_u8().ok_or(ParseError::UnexpectedEof)?;
        let kind = EventKind::from_byte(byte).unwrap_or_else(|| todo!("invalid event: {byte}"));

        match kind {
            EventKind::SyncBeep => {
                // TODO: SyncBeep handle time rollover

                let time = decode::variable(data)?;
                events.push(Self::SyncBeep(time.into()));
            }

            EventKind::InflightAdjustment => {
                let function = data.read_u8().ok_or(ParseError::UnexpectedEof)?;

                let new_value = if (function & 0x80) > 0 {
                    AdjustedValue::Float(data.read_f32().ok_or(ParseError::UnexpectedEof)?)
                } else {
                    AdjustedValue::Int(decode::variable_signed(data)?)
                };

                events.push(Self::InflightAdjustment {
                    function: function & 0x7F,
                    new_value,
                });
            }

            EventKind::Resume => {
                let log_iteration = decode::variable(data)?;
                let time = decode::variable(data)?;

                events.push(Self::Resume {
                    log_iteration,
                    time,
                });
            }

            EventKind::Disarm => {
                let reason = decode::variable(data)?;
                events.push(Self::Disarm(reason));
            }

            EventKind::FlightMode => {
                let flags = decode::variable(data)?;
                let last_flags = decode::variable(data)?;
                events.push(Self::FlightMode { flags, last_flags });
            }

            EventKind::ImuFailure => {
                let error = decode::variable(data)?;
                events.push(Self::ImuFailure { error });
            }

            EventKind::End => {
                check_message(data, b"End of log")?;

                if data.peek() == Some(b' ') {
                    // Assume INAV's new format:
                    // `End of log (disarm reason:x)\0`

                    check_message(data, b" (disarm reason:")?;

                    let reason = data.read_u8().ok_or(ParseError::UnexpectedEof)?;
                    events.push(Self::Disarm(reason.into()));

                    if data.read_u8() != Some(b')') {
                        return Err(ParseError::Corrupted);
                    }
                }

                if data.read_u8() != Some(0) {
                    return Err(ParseError::Corrupted);
                }

                events.push(Self::End);
            }
        }

        Ok(kind)
    }
}

byte_enum! {
    #[derive(Debug, Clone, Copy, PartialEq, Eq)]
    #[repr(u8)]
    pub(crate) enum EventKind {
        SyncBeep = 0,
        InflightAdjustment = 13,
        Resume = 14,
        Disarm = 15,
        FlightMode = 30,
        ImuFailure = 40,
        End = 255,
    }
}

fn check_message(bytes: &mut Reader, message: &[u8]) -> ParseResult<()> {
    let bytes = bytes.read_n_bytes(message.len());

    if bytes.len() != message.len() {
        return Err(ParseError::UnexpectedEof);
    }

    if bytes != message {
        return Err(ParseError::Corrupted);
    }

    Ok(())
}
