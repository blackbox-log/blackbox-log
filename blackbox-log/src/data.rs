//! Types for the data section of blackbox logs.

use alloc::vec::Vec;

use crate::event::{self, Event};
use crate::frame::{DataFrameKind, FrameKind, GpsFrame, GpsHomeFrame, MainFrame, SlowFrame};
use crate::parser::InternalError;
use crate::{Headers, Reader};

#[derive(Debug, Clone)]
pub(crate) struct Data {
    pub(crate) events: Vec<Event>,
    pub(crate) main_frames: Vec<FrameSync>,
    pub(crate) slow_frames: Vec<SlowFrame>,
    pub(crate) gps_frames: Vec<GpsFrame>,
    pub(crate) gps_home_frames: Vec<GpsHomeFrame>,
}

#[derive(Debug, Clone)]
pub(crate) struct FrameSync {
    pub(crate) main: MainFrame,
    pub(crate) slow: usize,
    pub(crate) gps: Option<usize>,
}

impl FrameSync {
    pub(crate) const fn new(main: MainFrame, slow: &[SlowFrame], gps: &[GpsFrame]) -> Self {
        Self {
            main,
            slow: slow.len() - 1,
            gps: gps.len().checked_sub(1),
        }
    }
}

/// Statistics about a decoded log.
#[derive(Debug, Clone)]
#[cfg_attr(feature = "serde", derive(serde::Serialize))]
#[non_exhaustive]
pub struct Stats {
    /// The number of valid frames found of each type.
    pub counts: FrameCounts,
}

#[derive(Debug, Clone)]
#[cfg_attr(feature = "serde", derive(serde::Serialize))]
pub struct FrameCounts {
    pub main: usize,
    pub slow: usize,
    pub gps: usize,
    pub gps_home: usize,
}

impl Data {
    pub(crate) fn to_stats(&self) -> Stats {
        Stats {
            counts: FrameCounts {
                main: self.main_frames.len(),
                slow: self.slow_frames.len(),
                gps: self.gps_frames.len(),
                gps_home: self.gps_home_frames.len(),
            },
        }
    }

    pub(crate) fn parse(data: &mut Reader, headers: &Headers) -> Self {
        let mut events = Vec::new();
        let mut main_frames = Vec::new();
        let mut slow_frames = Vec::new();
        let mut gps_frames = Vec::new();
        let mut gps_home_frames = Vec::new();

        slow_frames.push(headers.slow_frames.empty_frame());

        if let Some(def) = &headers.gps_frames {
            gps_frames.push(def.empty_frame());
        }

        let mut restore;
        let mut last_kind = None;
        while let Some(byte) = data.read_u8() {
            restore = data.get_restore_point();

            let Some(kind) = FrameKind::from_byte(byte) else {
                tracing::debug!("found invalid frame byte: {byte:0>#2x}");

                if let Some(last_kind) = last_kind.take() {
                    data.restore(restore);

                    match last_kind {
                        FrameKind::Event => {
                            events.pop();
                        }
                        FrameKind::Data(DataFrameKind::Intra | DataFrameKind::Inter) => {
                            main_frames.pop();
                        }
                        FrameKind::Data(DataFrameKind::Slow) => {
                            slow_frames.pop();
                        }
                        FrameKind::Data(DataFrameKind::Gps) => {
                            gps_frames.pop();
                        }
                        FrameKind::Data(DataFrameKind::GpsHome) => {
                            gps_home_frames.pop();
                        }
                    };
                }

                skip_to_frame(data);
                continue;
            };

            tracing::trace!("trying to parse {kind:?} frame");

            let result = match kind {
                FrameKind::Event => match Event::parse_into(data, &mut events) {
                    Ok(event::EventKind::End) => {
                        tracing::trace!("found the end event");
                        break;
                    }
                    Ok(_) => Ok(()),
                    Err(err) => Err(err),
                },
                FrameKind::Data(DataFrameKind::Intra | DataFrameKind::Inter) => {
                    let frame = MainFrame::parse(data, kind, &main_frames, headers);
                    frame.map(|frame| {
                        main_frames.push(FrameSync::new(frame, &slow_frames, &gps_frames));
                    })
                }
                FrameKind::Data(DataFrameKind::Gps) => headers.gps_frames.as_ref().map_or_else(
                    || {
                        tracing::debug!("found GPS frame without GPS frame definition");
                        Err(InternalError::Retry)
                    },
                    |gps| {
                        gps.parse(
                            data,
                            headers,
                            main_frames.last().map(|sync| &sync.main),
                            gps_home_frames.last(),
                        )
                        .map(|frame| gps_frames.push(frame))
                    },
                ),
                FrameKind::Data(DataFrameKind::GpsHome) => {
                    headers.gps_home_frames.as_ref().map_or_else(
                        || {
                            tracing::debug!(
                                "found GPS home frame without GPS home frame definition"
                            );
                            Err(InternalError::Retry)
                        },
                        |gps_home| {
                            gps_home
                                .parse(data, headers)
                                .map(|frame| gps_home_frames.push(frame))
                        },
                    )
                }
                FrameKind::Data(DataFrameKind::Slow) => headers
                    .slow_frames
                    .parse(data, headers)
                    .map(|frame| slow_frames.push(frame)),
            };

            match result {
                Ok(()) => {
                    last_kind = Some(kind);
                }
                Err(InternalError::Retry) => {
                    tracing::debug!("found corrupted {kind:?} frame");
                    data.restore(restore);
                    skip_to_frame(data);
                }
                Err(InternalError::Eof) => {
                    tracing::debug!("found unexpected end of file in data section");
                    break;
                }
            }
        }

        Self {
            events,
            main_frames,
            slow_frames,
            gps_frames,
            gps_home_frames,
        }
    }
}

#[cold]
fn skip_to_frame(data: &mut Reader) {
    data.skip_until_any(
        &[
            FrameKind::Event,
            FrameKind::Data(DataFrameKind::Intra),
            FrameKind::Data(DataFrameKind::Slow),
            FrameKind::Data(DataFrameKind::Gps),
            FrameKind::Data(DataFrameKind::GpsHome),
        ]
        .map(u8::from),
    );
}
