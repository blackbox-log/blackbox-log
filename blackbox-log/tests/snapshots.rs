use std::collections::BTreeMap;
use std::fs;
use std::io::Read;

use blackbox_log::data::{ParseEvent, Stats};
use blackbox_log::event::Event;
use blackbox_log::units::{si, Flag, FlagSet};
use blackbox_log::{DataParser, Headers, Unit, Value};
use serde::ser::{SerializeStruct, Serializer};
use serde::Serialize;

macro_rules! run {
    () => {
        |path| {
            let mut file = fs::File::open(path).unwrap();
            let mut data = Vec::new();
            file.read_to_end(&mut data).unwrap();

            let file = blackbox_log::File::new(&data);
            let logs = file
                .iter()
                .map(|mut reader| {
                    Headers::parse(&mut reader).map(|headers| {
                        let data = DataParser::new(&mut reader, &headers);
                        LogSnapshot::new(&headers, data)
                    })
                })
                .collect::<Vec<_>>();

            insta::assert_ron_snapshot!(logs);
        }
    };
}

#[test]
fn own() {
    insta::glob!("logs/*.bbl", run!());
}

#[test]
#[ignore]
fn fc_blackbox() {
    insta::glob!("logs/fc-blackbox/*", run!());
}

#[test]
#[ignore]
fn gimbal_ghost() {
    insta::glob!("logs/gimbal-ghost/*", run!());
}

#[derive(Debug, Serialize)]
struct LogSnapshot<'data> {
    headers: Headers<'data>,
    stats: Stats,
    events: Vec<Event>,
    main: Fields,
    slow: Fields,
    gps: Fields,
}

impl<'data> LogSnapshot<'data> {
    fn new(headers: &Headers<'data>, mut data: DataParser<'data, '_, '_>) -> Self {
        let headers = headers.clone();

        let mut events = Vec::new();
        let mut main = headers.main_def().iter().collect::<Fields>();
        let mut slow = headers.slow_def().iter().collect::<Fields>();
        let mut gps = headers
            .gps_def()
            .iter()
            .flat_map(|def| def.iter())
            .collect::<Fields>();

        while let Some(frame) = data.next() {
            match frame {
                ParseEvent::Event(event) => events.push(event),
                ParseEvent::Main(frame) => main.update(frame),
                ParseEvent::Slow(frame) => slow.update(frame),
                ParseEvent::Gps(frame) => gps.update(frame),
            }
        }

        Self {
            headers,
            stats: data.stats().clone(),
            events,
            main,
            slow,
            gps,
        }
    }
}

#[derive(Debug, Serialize)]
struct Fields(Vec<FieldSnapshot>);

impl Fields {
    fn update<F: blackbox_log::frame::Frame>(&mut self, frame: F) {
        for (field, value) in self.0.iter_mut().zip(frame.iter()) {
            field.update(value.into());
        }
    }
}

impl<T, U> FromIterator<(T, U)> for Fields
where
    T: Into<String>,
    U: Into<Unit>,
{
    fn from_iter<I: IntoIterator<Item = (T, U)>>(iter: I) -> Self {
        Self(
            iter.into_iter()
                .map(|(name, unit)| FieldSnapshot::new(name.into(), unit.into()))
                .collect(),
        )
    }
}

#[derive(Debug, Serialize)]
struct FieldSnapshot {
    name: String,
    unit: Unit,
    history: History,
}

#[derive(Debug, Serialize)]
#[serde(untagged)]
enum History {
    Int(NumberHistory<16, i128>),
    Bool { yes: usize, no: usize },
    Flags(BTreeMap<&'static str, usize>),
}

#[derive(Debug)]
struct NumberHistory<const N: usize, T> {
    min: T,
    max: T,
    seen: Vec<T>,
    histogram: [usize; N],
}

impl FieldSnapshot {
    fn new(name: String, unit: Unit) -> Self {
        Self {
            name,
            unit,
            history: match unit {
                Unit::FrameTime
                | Unit::Amperage
                | Unit::Voltage
                | Unit::Acceleration
                | Unit::Rotation
                | Unit::GpsCoordinate
                | Unit::Altitude
                | Unit::Velocity
                | Unit::GpsHeading
                | Unit::Unitless => History::Int(NumberHistory::new()),
                Unit::Boolean => History::Bool { yes: 0, no: 0 },
                Unit::FlightMode | Unit::State | Unit::FailsafePhase => {
                    History::Flags(BTreeMap::new())
                }
            },
        }
    }

    #[allow(clippy::wildcard_enum_match_arm, clippy::cast_possible_truncation)]
    fn update(&mut self, value: Value) {
        match &mut self.history {
            History::Int(history) => history.update(match value {
                Value::FrameTime(t) => t.get::<si::time::microsecond>().round() as i128,
                Value::Amperage(a) => a.get::<si::electric_current::milliampere>().round() as i128,
                Value::Voltage(v) => v.get::<si::electric_potential::millivolt>().round() as i128,
                Value::Acceleration(a) => a
                    .get::<si::acceleration::centimeter_per_second_squared>()
                    .round() as i128,
                Value::Rotation(r) => {
                    r.get::<si::angular_velocity::degree_per_second>().round() as i128
                }
                Value::GpsCoordinate(c) => (c * 10000000.).round() as i128,
                Value::Altitude(a) => a.get::<si::length::meter>().round() as i128,
                Value::Velocity(v) => {
                    v.get::<si::velocity::centimeter_per_second>().round() as i128
                }
                Value::GpsHeading(h) => (h * 10.).round() as i128,
                Value::Unsigned(u) => u.into(),
                Value::Signed(s) => s.into(),
                _ => unreachable!(),
            }),
            History::Bool { yes, no } => match value {
                Value::Boolean(true) => *yes += 1,
                Value::Boolean(false) => *no += 1,
                _ => unreachable!(),
            },
            History::Flags(history) => {
                if let Value::FailsafePhase(phase) = value {
                    *history.entry(phase.as_name()).or_insert(0) += 1;
                } else {
                    let flags = match value {
                        Value::FlightMode(m) => m.as_names(),
                        Value::State(s) => s.as_names(),
                        _ => unreachable!(),
                    };

                    for flag in flags {
                        *history.entry(flag).or_insert(0) += 1;
                    }
                }
            }
        }
    }
}

impl<const N: usize, T> NumberHistory<N, T> {
    fn new() -> Self
    where
        T: Default,
    {
        Self {
            min: T::default(),
            max: T::default(),
            seen: Vec::new(),
            histogram: [0; N],
        }
    }

    fn update_range(&mut self, value: T)
    where
        T: PartialOrd,
    {
        if value < self.min {
            self.min = value;
        } else if value > self.max {
            self.max = value;
        }
    }

    /// Store new value sorted in `seen` if not already present
    fn update_seen(&mut self, value: T)
    where
        T: PartialOrd,
    {
        let index = self.seen.partition_point(|x| *x < value);
        if self.seen.get(index) != Some(&value) {
            self.seen.insert(index, value);
        }
    }

    /// Bucket so that sequential values go in sequential buckets, since data is
    /// usually clustered
    #[inline(always)]
    fn update_histogram(&mut self, value: u128) {
        self.histogram[(value % N as u128) as usize] += 1;
    }
}

impl<const N: usize> NumberHistory<N, i128> {
    fn update(&mut self, value: i128) {
        self.update_range(value);
        self.update_seen(value);
        self.update_histogram(value.unsigned_abs());
    }
}

impl<const N: usize> Serialize for NumberHistory<N, i128>
where
    [usize; N]: Serialize,
{
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: Serializer,
    {
        serialize(self, self.min, self.max, serializer)
    }
}

impl<const N: usize> Serialize for NumberHistory<N, f64>
where
    [usize; N]: Serialize,
{
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: Serializer,
    {
        serialize(
            self,
            format!("{:.2}", self.min),
            format!("{:.2}", self.max),
            serializer,
        )
    }
}

#[inline(always)]
fn serialize<const N: usize, T, S, U>(
    snapshot: &NumberHistory<N, T>,
    min: U,
    max: U,
    serializer: S,
) -> Result<S::Ok, S::Error>
where
    S: Serializer,
    U: Serialize,
    [usize; N]: Serialize,
{
    let unique = snapshot.seen.len();

    let mut state = serializer.serialize_struct("FieldSnapshot", 4)?;

    state.serialize_field("min", &min)?;
    state.serialize_field("max", &max)?;
    state.serialize_field("unique", &unique)?;

    let histogram = "histogram";
    if unique > 1 {
        state.serialize_field(histogram, &snapshot.histogram)?;
    } else {
        state.skip_field(histogram)?;
    }

    state.end()
}
