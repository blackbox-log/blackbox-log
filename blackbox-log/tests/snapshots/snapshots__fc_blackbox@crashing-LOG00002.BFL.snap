---
source: blackbox-log/tests/snapshots.rs
expression: logs
input_file: blackbox-log/tests/logs/.fc-blackbox/src/test-data/crashing-LOG00002.BFL
---
[
  Ok(LogSnapshot(
    headers: Headers(
      version: V2,
      firmware_revision: "EmuFlight 3.7.0 (9366544) OMNIBUSF4SD",
      firmware_kind: Betaflight,
      board_info: None,
      craft_name: Some("25x25_blaxkbox_s"),
      vbat: None,
      current_meter: None,
      acceleration_1g: Some(2048),
      gyro_scale: Some(0.017453292),
      min_throttle: Some(1070),
      motor_output_range: Some(MotorOutputRange(1070, 2000)),
    ),
    stats: Stats(
      counts: FrameCounts(
        main: 171694,
        slow: 12,
        gps: 0,
        gps_home: 0,
      ),
    ),
    events: [
      End,
    ],
    fields: Fields([
      FieldSnapshot(
        name: "loopIteration",
        unit: Unitless,
        history: FieldSnapshot(
          min: 0,
          max: 692225,
          unique: 171694,
          histogram: (10734, 10734, 10733, 10732, 10732, 10732, 10732, 10731, 10730, 10730, 10729, 10729, 10729, 10729, 10729, 10729),
        ),
      ),
      FieldSnapshot(
        name: "time",
        unit: FrameTime,
        history: FieldSnapshot(
          min: 0,
          max: 106002824,
          unique: 171694,
          histogram: (10763, 10696, 10805, 10625, 10647, 10735, 10797, 10815, 10752, 10687, 10713, 10751, 10617, 10617, 10842, 10832),
        ),
      ),
      FieldSnapshot(
        name: "axisP[0]",
        unit: Unitless,
        history: FieldSnapshot(
          min: -1581,
          max: 364,
          unique: 1038,
          histogram: (15763, 15551, 46687, 16520, 11019, 9666, 8729, 7563, 6591, 5869, 5326, 5040, 4607, 4407, 4235, 4121),
        ),
      ),
      FieldSnapshot(
        name: "axisP[1]",
        unit: Unitless,
        history: FieldSnapshot(
          min: -1308,
          max: 1969,
          unique: 1400,
          histogram: (66190, 31165, 20318, 13740, 9805, 6840, 5037, 3876, 2988, 2459, 1965, 1746, 1456, 1431, 1430, 1248),
        ),
      ),
      FieldSnapshot(
        name: "axisP[2]",
        unit: Unitless,
        history: FieldSnapshot(
          min: -513,
          max: 580,
          unique: 1092,
          histogram: (16392, 14854, 14456, 40676, 28865, 11341, 8669, 6968, 5892, 4991, 4099, 3687, 3349, 2814, 2539, 2102),
        ),
      ),
      FieldSnapshot(
        name: "axisI[0]",
        unit: Unitless,
        history: FieldSnapshot(
          min: -336,
          max: 347,
          unique: 684,
          histogram: (53067, 8362, 6681, 8154, 6385, 7647, 8043, 8550, 7739, 7949, 8007, 9767, 10035, 8435, 6942, 5931),
        ),
      ),
      FieldSnapshot(
        name: "axisI[1]",
        unit: Unitless,
        history: FieldSnapshot(
          min: -46,
          max: 263,
          unique: 310,
          histogram: (62554, 16454, 10490, 6597, 4656, 6064, 5103, 6217, 6007, 6191, 6414, 5031, 6331, 8856, 6425, 8304),
        ),
      ),
      FieldSnapshot(
        name: "axisI[2]",
        unit: Unitless,
        history: FieldSnapshot(
          min: -243,
          max: 174,
          unique: 418,
          histogram: (53835, 9985, 10587, 8771, 8526, 7545, 6811, 5301, 8225, 8037, 6574, 10268, 6143, 7419, 8040, 5627),
        ),
      ),
      FieldSnapshot(
        name: "axisD[0]",
        unit: Unitless,
        history: FieldSnapshot(
          min: -1808,
          max: 1924,
          unique: 711,
          histogram: (60640, 20516, 17815, 15537, 12452, 10101, 8092, 6387, 5060, 3793, 3068, 2336, 1913, 1614, 1304, 1066),
        ),
      ),
      FieldSnapshot(
        name: "axisD[1]",
        unit: Unitless,
        history: FieldSnapshot(
          min: -3205,
          max: 2667,
          unique: 916,
          histogram: (68135, 26832, 20602, 15544, 11154, 8066, 5537, 3937, 2937, 2181, 1767, 1345, 1175, 956, 830, 696),
        ),
      ),
      FieldSnapshot(
        name: "axisD[2]",
        unit: Unitless,
        history: FieldSnapshot(
          min: -98,
          max: 104,
          unique: 131,
          histogram: (149627, 15042, 3532, 1155, 727, 435, 372, 219, 96, 105, 106, 67, 65, 55, 52, 39),
        ),
      ),
      FieldSnapshot(
        name: "axisF[0]",
        unit: Unitless,
        history: FieldSnapshot(
          min: 0,
          max: 0,
          unique: 1,
        ),
      ),
      FieldSnapshot(
        name: "axisF[1]",
        unit: Unitless,
        history: FieldSnapshot(
          min: 0,
          max: 0,
          unique: 1,
        ),
      ),
      FieldSnapshot(
        name: "axisF[2]",
        unit: Unitless,
        history: FieldSnapshot(
          min: 0,
          max: 0,
          unique: 1,
        ),
      ),
      FieldSnapshot(
        name: "rcCommand[0]",
        unit: Unitless,
        history: FieldSnapshot(
          min: -176,
          max: 87,
          unique: 264,
          histogram: (8577, 4835, 6800, 23467, 71505, 9955, 8449, 7636, 3935, 5431, 3975, 3710, 4840, 3302, 2635, 2642),
        ),
      ),
      FieldSnapshot(
        name: "rcCommand[1]",
        unit: Unitless,
        history: FieldSnapshot(
          min: -137,
          max: 94,
          unique: 232,
          histogram: (42309, 56799, 15042, 7744, 4394, 4666, 4647, 4110, 2634, 4123, 3700, 2871, 6132, 5430, 3046, 4047),
        ),
      ),
      FieldSnapshot(
        name: "rcCommand[2]",
        unit: Unitless,
        history: FieldSnapshot(
          min: -122,
          max: 267,
          unique: 389,
          histogram: (11338, 7444, 5393, 8378, 64058, 19235, 7160, 6369, 5923, 4491, 4582, 5148, 6443, 5381, 5554, 4797),
        ),
      ),
      FieldSnapshot(
        name: "rcCommand[3]",
        unit: Unitless,
        history: FieldSnapshot(
          min: 0,
          max: 1522,
          unique: 508,
          histogram: (4605, 8631, 4890, 4557, 5136, 3318, 4654, 60042, 7286, 5508, 2885, 4644, 15282, 23080, 9243, 7933),
        ),
      ),
      FieldSnapshot(
        name: "setpoint[0]",
        unit: Unitless,
        history: FieldSnapshot(
          min: -94,
          max: 40,
          unique: 135,
          histogram: (10894, 19176, 87333, 6001, 3076, 4453, 5048, 3651, 4277, 6259, 6339, 3160, 4039, 3677, 2102, 2209),
        ),
      ),
      FieldSnapshot(
        name: "setpoint[1]",
        unit: Unitless,
        history: FieldSnapshot(
          min: -68,
          max: 44,
          unique: 113,
          histogram: (89435, 19146, 7268, 6160, 2902, 6616, 5272, 6319, 3321, 3295, 3337, 4060, 3926, 3969, 2647, 4021),
        ),
      ),
      FieldSnapshot(
        name: "setpoint[2]",
        unit: Unitless,
        history: FieldSnapshot(
          min: -59,
          max: 171,
          unique: 231,
          histogram: (8192, 6969, 81091, 10348, 5272, 6942, 8964, 9083, 5873, 5297, 4517, 3571, 3682, 5120, 4008, 2765),
        ),
      ),
      FieldSnapshot(
        name: "setpoint[3]",
        unit: Unitless,
        history: FieldSnapshot(
          min: 0,
          max: 523,
          unique: 505,
          histogram: (63241, 4078, 4478, 3038, 14067, 8048, 23561, 8975, 6048, 5996, 7987, 3487, 5118, 4364, 3669, 5539),
        ),
      ),
      FieldSnapshot(
        name: "gyroADC[0]",
        unit: Rotation,
        history: FieldSnapshot(
          min: -229,
          max: 974,
          unique: 728,
          histogram: (51902, 24393, 18498, 14378, 11382, 8737, 7236, 6148, 5284, 4806, 4117, 3587, 2975, 2786, 2756, 2709),
        ),
      ),
      FieldSnapshot(
        name: "gyroADC[1]",
        unit: Rotation,
        history: FieldSnapshot(
          min: -1060,
          max: 703,
          unique: 892,
          histogram: (57725, 24857, 13695, 9552, 7783, 7245, 6602, 5542, 5079, 4637, 4795, 5504, 5672, 5313, 4354, 3339),
        ),
      ),
      FieldSnapshot(
        name: "gyroADC[2]",
        unit: Rotation,
        history: FieldSnapshot(
          min: -303,
          max: 263,
          unique: 567,
          histogram: (55719, 15588, 12189, 10992, 10355, 9267, 7543, 6117, 6386, 6551, 6864, 6498, 5005, 4437, 4144, 4039),
        ),
      ),
      FieldSnapshot(
        name: "accSmooth[0]",
        unit: Acceleration,
        history: FieldSnapshot(
          min: -10898,
          max: 6599,
          unique: 2060,
          histogram: (10627, 9419, 9628, 9637, 10325, 10227, 9918, 9222, 9028, 8824, 9651, 13033, 13290, 13663, 13583, 11619),
        ),
      ),
      FieldSnapshot(
        name: "accSmooth[1]",
        unit: Acceleration,
        history: FieldSnapshot(
          min: -8977,
          max: 5637,
          unique: 1802,
          histogram: (8473, 9314, 10109, 10104, 10498, 9921, 10551, 12692, 14579, 19364, 12692, 10166, 8654, 8440, 8067, 8070),
        ),
      ),
      FieldSnapshot(
        name: "accSmooth[2]",
        unit: Acceleration,
        history: FieldSnapshot(
          min: -10158,
          max: 10461,
          unique: 3113,
          histogram: (8586, 8262, 8860, 11342, 15400, 18144, 15652, 11748, 9775, 9497, 9650, 9631, 9266, 8759, 8586, 8536),
        ),
      ),
      FieldSnapshot(
        name: "motor[0]",
        unit: Unitless,
        history: FieldSnapshot(
          min: 0,
          max: 1980,
          unique: 879,
          histogram: (8514, 8364, 8278, 8067, 8009, 8138, 7838, 7544, 7384, 7481, 7356, 7483, 7792, 11534, 48974, 8938),
        ),
      ),
      FieldSnapshot(
        name: "motor[1]",
        unit: Unitless,
        history: FieldSnapshot(
          min: 0,
          max: 4294967295,
          unique: 717,
          histogram: (59941, 5599, 5606, 5668, 6664, 14012, 19520, 11122, 6227, 5423, 5445, 5515, 5384, 5145, 5161, 5262),
        ),
      ),
      FieldSnapshot(
        name: "motor[2]",
        unit: Unitless,
        history: FieldSnapshot(
          min: 0,
          max: 4294967295,
          unique: 774,
          histogram: (45070, 18364, 7909, 7679, 7795, 7909, 8216, 7796, 7869, 7712, 7614, 7723, 8838, 7428, 6952, 6820),
        ),
      ),
      FieldSnapshot(
        name: "motor[3]",
        unit: Unitless,
        history: FieldSnapshot(
          min: 0,
          max: 4294967295,
          unique: 583,
          histogram: (91388, 6039, 5676, 5375, 5368, 5402, 5489, 5246, 5384, 5233, 5043, 5058, 5132, 5099, 5408, 5354),
        ),
      ),
      FieldSnapshot(
        name: "flightModeFlags",
        unit: FlightMode,
        history: {
          "Arm": 171693,
          "ThreeD": 171693,
        },
      ),
      FieldSnapshot(
        name: "stateFlags",
        unit: State,
        history: {},
      ),
      FieldSnapshot(
        name: "failsafePhase",
        unit: FailsafePhase,
        history: {},
      ),
      FieldSnapshot(
        name: "rxSignalReceived",
        unit: Boolean,
        history: History(
          yes: 171693,
          no: 0,
        ),
      ),
      FieldSnapshot(
        name: "rxFlightChannelsValid",
        unit: Boolean,
        history: History(
          yes: 171693,
          no: 0,
        ),
      ),
    ]),
  )),
]
